### ISSUE TITLE
Native library (com/sun/jna/darwin/libjnidispatch.jnilib) not found in resource path

### ISSUE THREAD

**Issue by avierr** [oracle/graal#1332]
Not sure, if this is the right place for this issue, I am currently using micronaut 1.1.2 to build the binary.
The binary builds perfectly, it runs well. 
but it throws the following exception from within a scheduled task. 

Not sure if this is a problem, but this is shown when the binary first loads:
```
WARNING: The sunec native library, required by the SunEC provider, could not be loaded. This library is usually shipped as part of the JDK and can be found under <JAVA_HOME>/jre/lib/<platform>/libsunec.so. It is loaded at run time via System.loadLibrary("sunec"), the first time services from SunEC are accessed. To use this provider's services the java.library.path system property needs to be set accordingly to point to a location that contains libsunec.so. Note that if java.library.path is not set it defaults to the current working directory.
```

The error:
```
12:50:47.305 [pool-2-thread-4] ERROR i.m.s.DefaultTaskExceptionHandler - Error invoking scheduled task for bean [rdf.agent.supervisor.tasks.Supervisor@118320cf8] Native library (com/sun/jna/darwin/libjnidispatch.jnilib) not found in resource path ()
java.lang.UnsatisfiedLinkError: Native library (com/sun/jna/darwin/libjnidispatch.jnilib) not found in resource path ()
        at com.sun.jna.Native.loadNativeDispatchLibraryFromClasspath(Native.java:1026)
        at com.sun.jna.Native.loadNativeDispatchLibrary(Native.java:988)
        at com.sun.jna.Native.<clinit>(Native.java:195)
        at com.oracle.svm.core.hub.ClassInitializationInfo.invokeClassInitializer(ClassInitializationInfo.java:347)
        at com.oracle.svm.core.hub.ClassInitializationInfo.initialize(ClassInitializationInfo.java:267)
        at java.lang.Class.ensureInitialized(DynamicHub.java:437)
        at com.sun.jna.Memory.malloc(Memory.java:721)
        at com.sun.jna.Memory.<init>(Memory.java:113)
        at com.sun.jna.ptr.ByReference.<init>(ByReference.java:42)
        at com.sun.jna.ptr.IntByReference.<init>(IntByReference.java:33)
        at com.sun.jna.ptr.IntByReference.<init>(IntByReference.java:29)
        at oshi.util.platform.mac.SysctlUtil.sysctl(SysctlUtil.java:99)
        at oshi.software.os.mac.MacOSVersionInfoEx.<init>(MacOSVersionInfoEx.java:44)
        at oshi.software.os.mac.MacOperatingSystem.<init>(MacOperatingSystem.java:74)
        at oshi.SystemInfo.getOperatingSystem(SystemInfo.java:103)
        at rdf.agent.supervisor.tasks.Supervisor.isAgentAlive(Supervisor.java:80)
        at rdf.agent.supervisor.tasks.Supervisor.checkAgent(Supervisor.java:60)
        at rdf.agent.supervisor.tasks.$SupervisorDefinition$$exec1.invokeInternal(Unknown Source)
        at io.micronaut.context.AbstractExecutableMethod.invoke(AbstractExecutableMethod.java:144)
        at io.micronaut.inject.DelegatingExecutableMethod.invoke(DelegatingExecutableMethod.java:76)
        at io.micronaut.scheduling.processor.ScheduledMethodProcessor.lambda$process$5(ScheduledMethodProcessor.java:110)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:473)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:193)

```

I hope there is a magical switch somewhere as this is the only thing that's blocking me from finally getting a proper native binary to work on a production environment.

**Comment by christianwimmer** [comment]
We have never tried JNA inside a native image, and I don't know the internals how JNA works. If it is based on Java bytecode generation, then all the bytecode generation needs to be done at image generation time.

**Comment by peter-hofer** [comment]
Your scheduled task calls into the _oshi_ library that uses Java Native Access (JNA), which tries to load a native helper library as a Java resource. In order to include the library file in the image, you will need to add an entry in the resource configuration file as described in [RESOURCES.md](https://ol-bitbucket.us.oracle.com/projects/G/repos/graal/browse/substratevm/RESOURCES.md). `.*/libjnidispatch.*` as pattern should work. You can also try to work around this for now by extracting the library file from its JNA JAR file (presumably) and setting the `jna.boot.library.path` property, according to [jna/Native.java](https://github.com/java-native-access/jna/blob/master/src/com/sun/jna/Native.java#L916).

However, JNA in general might not work with native-image unless you can avoid code generation and class loading, see #673. If your application in general runs well, it could be worth looking into avoiding this specific dependency on oshi and JNA.

**Comment by jvican** [comment]
@peter-hofer I ended up following both of your workaroudns for a minimal application accessing `isatty` via JNA. I also had to tweak the proxy configuration, but that's not relevant for now. Before any proxy is loaded, it looks like JNA crashes in this line https://github.com/java-native-access/jna/blob/044cd6f0c173ade94269ce4b2742028ddfc0d4cd/src/com/sun/jna/Native.java#L1018
  
The issue can be reproduced by copying the `libjnidispatch.jnilib` file from the java-native-access/jna GitHub repo to a temporary file and then manually loading it with `System.load`. The issue seems to be that the native application shows me a popup asking me to install Java:

<img width="532" alt="Screenshot 2019-06-24 at 16 39 46" src="https://user-images.githubusercontent.com/2462974/60027946-c247df00-969e-11e9-834a-b6706097c3ed.png">

And then exist when it sees that no Java is installed.
  
<img width="1401" alt="Screenshot 2019-06-24 at 16 39 58" src="https://user-images.githubusercontent.com/2462974/60027978-cecc3780-969e-11e9-9947-7bc069849b17.png">

I'm not sure what could be going on here. Is there something I need to make that `System.load` call succeed?
  
@avierr FTR, did you run into this issue too?
